### PR TITLE
GaussianBlur and ColorJitter ops.

### PR DESCRIPTION
**Pull request by barucden**
This PR introduces first two non-geometric operations, Gaussian blur and contrast & brightness adjustment. It *partially* solves #16.

At first, I thought I could implement it easily but I encountered some issues. I would greatly appreciate help resolving them (cc @johnnychen94).

1. The introduced operations change the image eltype (e.g., `ColorTypes.RGBA{FixedPointNumbers.N0f8}` => `ColorTypes.RGBA{Float64}`). Is this something we need to worry about? This concern will perhaps apply for all non-geometric operations in future.
2. What is a good way to test these operations? It is difficult to come up with a test-case whose expected output is not based on the formulas present in the code (e.g., for `Rotate90` we know the expected output without evaluating the operation, which is not the case for `GaussianBlur`).
3. For a segmentation task, these operations should be applied only on the input image, not the ground truth segmentation. How can we deal with that? Should we introduce a new abstract subtype for this purpose?

I have two additional issues regarding the `AdjustContrastBrightness`:

1. The name seems too long. What would be a shorter alternative?
2. The formula is now `out[i] = clamp(α * img[i] + β * M)`, where `M=mean(img)`. However, sometimes it is better to use `M=the maximum value` (e.g., for a grayscale image whose pixels are represented as 8-bit unsigned integer, M=255). I could not find a generic way how to get this value for different color types (yet, I am not too experienced). Is there a way?

### CODE REVIEW

**Comment by johnnychen94** [comment]
> The introduced operations change the image eltype (e.g., ColorTypes.RGBA{FixedPointNumbers.N0f8} => ColorTypes.RGBA{Float64}). Is this something we need to worry about?

This is a separate issue. Most algorithms in JuliaImages don't take this assumption seriously so I think we can live this type of promotion without worrying about it too much. Generally, this issue can be fixed in upstream packages (e.g., ColorVectorSpace, FixedPointNumbers).

> What is a good way to test these operations?

I would just use something as simple as `ref = imfilter(img, gaussian((σ, σ), (k, k)))` to be the test reference, but wrap it with `assess_psnr`(from ImageQualityIndexes) to allow some internal changes. This test looks trivial but it provides some guarantee that internal `applyeager` is correctly called.

> For a segmentation task, these operations should be applied only on the input image, not the ground truth segmentation.

This is also a separate issue, maybe you can open a new issue with a minimal working example to discuss it? `augment` supports two usages:

```julia
augment(input_img, pipeline)
augment((input_img, ground_truth), pipeline) # augmentation applies uniformly to both images
```

I assume that you can just use the first version.

---

> I have two additional issues regarding the AdjustContrastBrightness:

I forgot to mention this, maybe we can directly use https://github.com/JuliaImages/ImageContrastAdjustment.jl for the eager version, it introduces more contrast adjustment methods, including `LinearStretching`? Then open an issue in that repo for a lazy view. How do you think of this?

I would like to see more functionality of Augmentor being implemented in other packages, and leave Augmentor a pipeline composer.

File: src/operations/blur.jl
Comment (johnnychen94) [inline src/operations/blur.jl]: See also https://juliaimages.org/ImageFiltering.jl/dev/#Factored-kernels for a more efficient version of this; two 1-d gaussian blur is equivalent to one 2-d gaussian blur.

Comment (barucden) [reply src/operations/blur.jl]: I am using `KernelFactors.gaussian`, so eventually [this method](https://github.com/JuliaImages/ImageFiltering.jl/blob/f85e7b0ad4cd6fb8499ccdf25712e6f00be5b982/src/kernelfactors.jl#L427) gets invoked. I think it is what you are proposing, isn't it?

**Comment by barucden** [comment]
I added tests for `GaussianBlur` but I use the standard pixel-wise comparison to check `applyeager`. I was trying to use `assess_psnr` but it seems that it lacks an implementation for RGB**A** images?

I like your idea about using ImageContrastAdjustment.jl. I am just not clear about the next step. Should I remove the proposed operation `AdjustContrastBrightness` for now? Or, as it seems that ImageContrastAdjustment.jl does not provide an operation equivalent to `AdjustContrastBrightness`, should I replace it with a different operation (such as `LinearStretching`) that comes from ImageContrastAdjustment.jl?

File: src/operations/brightness.jl
Comment (johnnychen94) [inline src/operations/brightness.jl]: It is more self-explaining if you can add the math `clamp(α * img[i] + β * M)` here.

File: src/operations/blur.jl
Comment (johnnychen94) [reply src/operations/blur.jl]: Yeah, you're right, I missed that `using` statement.

**Review by johnnychen94 — COMMENTED** [review]
> I was trying to use `assess_psnr` but it seems that it lacks an implementation for RGBA images?

Could possibly just use `RGB.(testpattern())` as image input.

It's not very clear to me how should we process the alpha channel of RGBA. Are there any references/conventions for this?

> However, sometimes it is better to use M=the maximum value (e.g., for a grayscale image whose pixels are represented as 8-bit unsigned integer, M=255). I could not find a generic way how to get this value for different color types (yet, I am not too experienced). Is there a way?

`gamutmin`/`gamutmax` from ColorTypes, and is also available in ImageCore (via Reexport).

> I like your idea about using ImageContrastAdjustment.jl. I am just not clear about the next step. Should I remove the proposed operation AdjustContrastBrightness for now? Or, as it seems that ImageContrastAdjustment.jl does not provide an operation equivalent to AdjustContrastBrightness, should I replace it with a different operation (such as LinearStretching) that comes from ImageContrastAdjustment.jl?

Ah never mind. I think this is for different purposes. Although `clamp(α * img[i] + β * M)` is mathematically a linear stretching operation, it seems a more convenient way for augmentation purposes; while the ImageContrastAdjustment `LinearStretching` output is more human-readable (e.g., stretch to [0, 1] range).

---

The implementation looks good to me, let me know when it is ready.

**Comment by johnnychen94** [comment]
One thing I forgot to mention, can you also add two cards to `docs/operations`? You can use other cards as templates and insert the filename to the corresponding `config.json`.

File: src/operations/color.jl
Comment (barucden) [inline src/operations/color.jl]: Is this OK?

**Comment by barucden** [comment]
new changes:
- changed the name from `AdjustContrastBrightness` to `ColorJitter` (torchvision uses this name for a similar method)
- added the feature based on `gamutmax`
- added cards to `docs/`

Regarding the PSNR computed for RGBA images, I am not sure. I tried searching, but did not find any resources. Maybe the alpha channel could be treated just like any other channel?

I think that we can settle with converting to RGB. What is a reasonable PSNR threshold for which two images are considered equal?

**Comment by johnnychen94** [comment]
> Maybe the alpha channel could be treated just like any other channel?

For JuliaImages that ranges consistently from [0,1] this is reasonable. But for other library that uses [0, 255] range, it doesn't make much sense to me. Anyway, we doesn't need to worry about this in this PR.

> What is a reasonable PSNR threshold for which two images are considered equal?

larger than 25-30 should be okay. I use 25 in ReferenceTests.

File: src/operations/color.jl
Comment (johnnychen94) [reply src/operations/color.jl]: Looks good to me. But to keep [type stability](https://docs.julialang.org/en/v1/manual/performance-tips/#Write-%22type-stable%22-functions), please also use `T(mean(img))`.

**Comment by barucden** [comment]
Okay, I think it is ready now. Thanks for your advices @johnnychen94! Let me know if I should change something.

File: test/operations/tst_color.jl
Comment (johnnychen94) [inline test/operations/tst_color.jl]: Is it `usemax=true`?

```suggestion
            res = Augmentor.applylazy(ColorJitter(α, β, usemax=true), img)
```

File: test/runtests.jl
Comment (johnnychen94) [inline test/runtests.jl:55]: Looks like "operations/tst_color.jl" is missing here.

File: src/operations/color.jl
Comment (johnnychen94) [inline src/operations/color.jl:59]: Is there any reference for this default value? Looks like torchvision's default value provides a no-op?

File: src/operations/color.jl
Comment (johnnychen94) [inline src/operations/color.jl]: This passes outer `@inbounds` macro to internal `@inbounds A.orig[i, j]`.

```suggestion
Base.@propagate_inbounds Base.getindex(A::ColorJitterView, i, j) = _map_pix(A.α, A.β, A.M, A.orig[i, j])
```

File: src/operations/blur.jl
Comment (johnnychen94) [inline src/operations/blur.jl:52]: Similarly, is there any reference for this default value? If so, please leave a comment here so that future maintainer knows what happens here.

File: src/operations/blur.jl
Comment (johnnychen94) [inline src/operations/blur.jl]: Just in case someone really passes a 1-D or 3D image here:

```suggestion
function applyeager(op::GaussianBlur, img::AbstractArray, (k, σ))
    n = ndims(img)
    kernel = gaussian(ntuple(_->σ, n), ntuple(_->k, n))
    return imfilter(img, kernel)
end
```

**Review by johnnychen94 — APPROVED** [review]
Only some small suggestions and questions. (It's midnight here, I'll come back tomorrow and approve tests if there are new commits; Github action by default disable the test running for non-contributor because months ago there are people sending spam PRs to other repositories to run mining script)

File: src/operations/blur.jl
Comment (barucden) [reply src/operations/blur.jl:52]: Both come from Albumentations -- note added.

File: test/runtests.jl
Comment (barucden) [reply test/runtests.jl:55]: Thanks for the catch. Turned out there were still errors in `_get_M` which the test caught after I included it in the list :smile: I pushed a fix together with a test for the `_get_M` method.

**Comment by johnnychen94** [comment]
Really nice work! The detailed descriptions in this PR and other issues are really appreciated.

**Pull request merged**
